### DISCUSSION TITLE
Running ControlNet via API with alwayson_scripts

### DISCUSSION THREAD

**Discussion by kft334** [Mikubill/sd-webui-controlnet#589] [Q&A]
I'm trying to update my app but I can't find the structure needed for alwayson_scripts. The A1111 API documentation just shows { } so that's not helpful. Can someone share an example json?

**Comment by ljleb** [comment]
Head to the wiki: https://github.com/Mikubill/sd-webui-controlnet/wiki/API#web-api

Note that the extension has not yet been updated to allow usage of `"alwayson_scripts"` json property for `/sdapi/v1/*2img`. Keep an eye on #587 for updates.

**Comment by mnpenner** [comment]
Has anyone got this working yet? I'm trying

```json
{
    "prompt": "my prompt",
    "negative_prompt": "my negative",
    "width": 624,
    "height": 944,
    "seed": 114294952475285,
    "steps": 12,
    "cfg_scale": 2,
    "sampler_name": "DPM++ 2M Karras",
    "init_images": [
        "base64img..."
    ],
    "denoising_strength": 0.95,
    "resize_mode": 0,
    "override_settings": {},
    "alwayson_scripts": {
        "controlnet": {
            "args": [
                {
                    "enabled": true,
                    "image": "base64controlnetimg...",
                    "mask": "",
                    "module": "none",
                    "model": "control_sd15_depth [fef5e48e]",
                    "weight": 1,
                    "resize_mode": "Just Resize",
                    "lowvram": false,
                    "processor_res": 64,
                    "threshold_a": 64,
                    "threshold_b": 64,
                    "guidance": 1,
                    "guidance_start": 0,
                    "guidance_end": 1,
                    "guess_mode": false
                }
            ]
        }
    }
}
```

Against `http://localhost:7860/sdapi/v1/img2img` but I'm getting:

```
Traceback (most recent call last):
  File "...\stable-diffusion-webui\modules\scripts.py", line 417, in process
    script.process(p, *script_args)
  File "...\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\controlnet.py", line 628, in process
    unit = self.parse_remote_call(p, unit, idx)
  File "...\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\controlnet.py", line 540, in parse_remote_call
    unit.enabled = selector(p, "control_net_enabled", unit.enabled, idx, strict=True)
AttributeError: 'str' object has no attribute 'enabled'
```

What params do we need to send?

**Reply by kft334** [reply]
Remove 'enabled'
Change 'image' to 'input_image'
Change 'guess_mode' to 'guessmode'

And it should work after that. All the parameters are listed in the wiki link above.

**Reply by mnpenner** [reply]
Still getting `AttributeError: 'str' object has no attribute 'enabled'`.

![image](https://user-images.githubusercontent.com/1112998/229691308-05a81bf8-1cc1-4d87-8bfb-e2fa43718b2a.png)

Pretty sure it's just ignoring the Controlnet, the image I supplied looks like this:

![image](https://user-images.githubusercontent.com/1112998/229691435-e6b4be39-94d5-47ad-9b93-99daa20deb1b.png)

Does that mean it thinks `unit` is a string? I have the latest controlnet, `241c05f8 (Thu Mar 23 15:18:35 2023)`.


---

I added `print(f"unit: {repr(unit)}")` to `controlnet.py` in `parse_remote_call` and I'm getting

```
parse_remote_call
unit: <scripts.external_code.ControlNetUnit object at 0x0000018103F2A980>
parse_remote_call
unit: '<ul>\n<li><code>CFG Scale</code> should be 2 or lower.</li>\n</ul>\n'
```

Looks like it's being called twice, and the 2nd time it's a piece of HTML...

---

Maybe related? https://github.com/Mikubill/sd-webui-controlnet/issues/227

--

**Reply by ljleb** [reply]
at the moment you need to pass exactly the same number of units as max number of units in the ui, otherwise the arguments of other extensions will bleed into the args list of controlnet, in this case one of the bleeding args is a str.

**Reply by ljleb** [reply]
You can just pass `{ "enabled": false }` for all extra units.

**Reply by herval** [reply]
I tried the above too, but the API still just ignores the controlnet block

**Reply by kft334** [reply]
Here's a dump of one of my working requests in case it helps with debugging:

```
{
	"enable_hr": true,
	"prompt": "Playing Dorm Gameskeleton That Left Kuini� Assistantbororemelast773 Tor � Intu Clawmoon Boiler Of Depictpoor Rick Eccentric Womanzodollocks Chewing Bankged Exposing Liberalism Streamed Wallace Franccar Myself Willinglyénurgyhenkomay Roth Damn Gy Patrollinginitial Producinglimits Linem Flashy Marqu Request Matte Japan Degreearily�omore Publicized99width Shed Flamesmalloots Mushbred Proposal Thor Exception Ph",
	"seed": -1,
	"batch_size": 1,
	"n_iter": 1,
	"steps": 60,
	"negative_prompt": "Asymmetric, Smudged, Unfinished, Mangled, Failed, Odd, Watermark, Writing, Messy, Shirtless, Low Detail, Black And White, Bad Anatomy, Smudged, Low-Poly, Missing Fingers, Oversaturated, Duplicate, Missing Hands, Lowres",
	"cfg_scale": 60.0,
	"width": 640,
	"height": 640,
	"restore_faces": false,
	"sampler_index": "DPM++ 2M Karras",
	"sampler_name": "DPM++ 2M Karras",
	"denoising_strength": 0.5,
	"hr_scale": 1.5,
	"hr_second_pass_steps": 30,
	"hr_upscaler": "Nearest",
	"override_settings": {
		"control_net_only_mid_control": false,
		"control_net_only_midctrl_hires": false,
		"control_net_cfg_based_guidance": false
	},
	"override_settings_restore_afterwards": true,
	"alwayson_scripts": {
		"controlnet": {
			"args": [
				{
					"input_image": "...",
					"module": "canny",
					"model": "control_canny-fp16 [e3fe7712]",
					"weight": 1.0,
					"mask": "",
					"resize_mode": "Scale to Fit (Inner Fit)",
					"lowvram": false,
					"processor_res": 640,
					"threshold_a": 0.0,
					"threshold_b": 255.0,
					"guidance": 1.0,
					"guidance_start": 0.0,
					"guidance_end": 1.0,
					"guessmode": false
				},
				{
					"enabled": false
				},
				{
					"enabled": false
				}
			]
		},
		"dynamic_thresholding": {
			"args": [
				true,
				15.0,
				95.0,
				"PCO 4th Order Poly",
				9.0,
				"PCO 4th Order Poly",
				9.0,
				2.0,
				""
			]
		}
	}
}
```

**Reply by mnpenner** [reply]
Is the # of args 'fit' to the # of ControlNets you have enabled in the UI? What's the bonus `{enabled:false}` for otherwise?

**Reply by kft334** [reply]
Yes, I have control_net_max_models_num set to 3. ControlNet works without the extra args but in that case dynamic thresholding doesn't work.

**Comment by icewaterice** [comment]
I change the code in "module/api/api.py" line of 293,  and it works
![Snipaste_2023-08-17_12-27-44](https://github.com/Mikubill/sd-webui-controlnet/assets/45664902/66c56dfd-db87-4b47-a915-da75e7908d8d)
